Fixed IDE services touching `unmanaged` constraint

OmarTawfik opened the pull request:
In this PR:

* Completion Service
* Extract Interface Service
* Implement interface service
* Metadata As Source Service
* Upgrade Project Service
* Quick Info and Symbol Display
* Classification Service

Each service has the corresponding tests

Document: https://github.com/dotnet/csharplang/blob/master/proposals/blittable.md

OmarTawfik commented:
cc @dotnet/compiler-api for API addition
cc @dotnet/roslyn-compiler for small Symbol Display change
cc @dotnet/roslyn-ide for code review.

CyrusNajmabadi commented:
Can we have a test for GenerateMethod (where it captures a value of "T:unmaanged")?

OmarTawfik commented:
@CyrusNajmabadi
Added the required tests, and added a test for finding references when it is a type. But when it is not, there is no symbol to use in the find references API. is there a test that does so for a given source location to assert that no results are returned?

OmarTawfik commented:
cc @dotnet/compiler-api for API addition
cc @dotnet/roslyn-compiler for small Symbol Display change
cc @dotnet/roslyn-ide for code review

jaredpar commented:
CC @roslyn-ide for review

jaredpar commented:
API approved

sharwell commented on src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs:
```diff
@@ -2325,5 +2325,118 @@ void Main(string s)
+        {
+            await TestInMethodAsync(
+                "var unmanaged = 0;",
+                Keyword("var"));
```
❓ Does this automatically verify that `unmanaged` is *not* marked as a keyword here?

OmarTawfik replied:
Yes. There should've been another entry if it was a keyword. I'll add a comment.

---
In reply to: [171661262](https://github.com/dotnet/roslyn/pull/25033#discussion_r171661262) [](ancestors = 171661262)

sharwell commented on src/EditorFeatures/CSharpTest/Classification/TotalClassifierTests.cs:
```diff
@@ -849,5 +849,263 @@ class Program : IReadOnlyCollection<int,string>
+        public async Task TestUnmanagedConstraint_InsideMethod()
+        {
+            await TestInMethodAsync(
+                "var unmanaged = 0;",
```
❗️ Need tests for accesses to `unmanaged` as well

sharwell commented on src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs:
```diff
@@ -472,5 +472,42 @@ void A()
+        {
+            await TestLanguageVersionUpgradedAsync(
+@"
+class Test<T> where T : [|unmanaged|]
```
💡 Also test that this is not offered if `unmanaged` is already defined as an interface

jasonmalinowski commented on src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateMethodTests.cs:
```diff
@@ -4165,42 +4170,47 @@ private static void Bar<T>(Func<List<T>> p)
 }");
         }
 
+        [Theory]
```
Nice, I like this!

jasonmalinowski commented on src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:
```diff
@@ -146,6 +146,22 @@ private static bool IsNamespaceName(NameSyntax name)
                 }
             }
 
+            if (name.IsUnmanaged && name.Parent.IsKind(SyntaxKind.TypeConstraint))
```
Paging @CyrusNajmabadi on this little bit.

CyrusNajmabadi replied:
This looks correct to me.  

Note: i don't like the .IsUnmanaged property as it sounds so ambiguous.  But it's parallel to the .IsVar property we have... so i guess we should keep it :)

jasonmalinowski commented on src/Workspaces/CSharp/Portable/Extensions/ITypeParameterSymbolExtensions.cs:
```diff
@@ -42,6 +42,10 @@ internal static class ITypeParameterSymbolExtensions
             }
+            else if (typeParameter.HasUnmanagedTypeConstraint)
+            {
+                constraints.Add(SyntaxFactory.TypeConstraint(SyntaxFactory.IdentifierName("unmanaged")));
```
Should this be SyntaxFactory.ClassOrStruct constraint? Or should there be a new syntax factory method? I guess I think of this as analogous to the keyword ones, hence wondering if it fits there.

CyrusNajmabadi replied:
It should generate the same code as what was parsed out IMO.  So this should be a "TypeConstraint" with "unmanaged" as the type name.  just like "var" is parsed out just an type with an identifier name.  It's only semantic analysis that imbues these identifiers with special meaning.

sharwell commented on src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs:
```diff
@@ -279,14 +295,14 @@ private static ISymbol TryGetSymbol(NameSyntax name, SymbolInfo symbolInfo, Sema
             }
 
-            return symbol != null && symbol.Name == "var";
+            return symbol != null && symbol.Name == name;
```
💭 Seems like the type/arity check could come last since the name check is the most likely to fail here.

OmarTawfik commented:
@jasonmalinowski @sharwell comments addressed

a couple of notes about cross consumption: VB doesn't support this feature, so the VB compiler will reject any symbols that have the constraint. Please check tests merged in an earlier PR in:

https://github.com/dotnet/roslyn/blob/features/constraints/src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadCustomModifiers.vb

You will not be able to use the symbols, instantiate such types, or call into such methods.
I'm not sure what to test here on the IDE side regarding poisoned symbols.

OmarTawfik commented:
@jasonmalinowski are there any other comments? I'd like to merge this Tuesday if possible.

jasonmalinowski commented:
@OmarTawfik Can we get a test or two that the rejection happens how we anticipate? When you say "you will not be able to us the symbols" is it just that the compiler won't have symbols from metadata at all (i.e. the type is invisible) or that the ISymbols from the compiler are somehow toxic and will explode when touched?

jasonmalinowski commented:
@OmarTawfik Or perhaps put another way: where's the spec that says what those unsupported symbols will be imported?

OmarTawfik commented:
@jasonmalinowski
For VB, there is no spec because the feature is not supported. To "unsupport" it in VB we use a modreq over the type parameter constraint. It is mentioned in the spec.
The link I posted in my previous comment assets how would VB react to such modreq. It produces errors when referencing or invoking such symbols.
As an errornous symbol, VB will do its best and treat it like a `struct` constraint anyways.

OmarTawfik commented:
@jasonmalinowski modified the VB tests to also assert that the previous constraint I mentioned holds. Please let me know if there are any other concerns.

jasonmalinowski commented:
@OmarTawfik Thanks! My concern is as far as the IDE is concerned "X isn't supported" is our normal operating state. The syntax trees and symbols we use are almost always incomplete; if the symbols are just disappearing then we'll end up with strange behaviors and possible crashes.
